From d393f7c9174a4abe64dbba364e3b42bb20768183 Mon Sep 17 00:00:00 2001
From: Bruce Schmid <duck@freescale.com>
Date: Fri, 6 Jun 2008 15:18:16 -0600
Subject: [PATCH] ENGR00083330-7 Integrate Devtech's USB code patch 7/12

USB: mx27 gpio cleanups.
Return status from USB pin grab routines.  Don't initialize statics.

Signed-off-by: Bruce Schmid <duck@freescale.com>
---
 arch/arm/mach-mx27/mx27ads_gpio.c |  129 +++++++++++++++++++-----------------
 1 files changed, 68 insertions(+), 61 deletions(-)

diff --git a/arch/arm/mach-mx27/mx27ads_gpio.c b/arch/arm/mach-mx27/mx27ads_gpio.c
index a6aa782..aaf62c0 100644
--- a/arch/arm/mach-mx27/mx27ads_gpio.c
+++ b/arch/arm/mach-mx27/mx27ads_gpio.c
@@ -175,7 +175,7 @@ void config_uartdma_event(int port)
 	return;
 }
 
-static int usbh1_hs_active = 0;
+static int usbh1_hs_active;
 /*!
  * Setup GPIO for USB, Total 34 signals
  * PIN Configuration for USBOTG:   High/Full speed OTG
@@ -191,23 +191,25 @@ static int usbh1_hs_active = 0;
   *  PB25 - PB31  -- PRIMARY
       PB22  -- PRIMARY
  */
-void gpio_usbh1_active(void)
+int gpio_usbh1_active(void)
 {
 	if (usbh1_hs_active)
-		return;
-	gpio_request_mux(MX27_PIN_USBH1_SUSP, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBH1_RCV, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBH1_FS, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBH1_OE_B, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBH1_TXDM, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBH1_TXDP, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBH1_RXDM, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBH1_RXDP, GPIO_MUX_PRIMARY);
+		return 0;
+
+	if (gpio_request_mux(MX27_PIN_USBH1_SUSP, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBH1_RCV, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBH1_FS, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBH1_OE_B, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBH1_TXDM, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBH1_TXDP, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBH1_RXDM, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBH1_RXDP, GPIO_MUX_PRIMARY))
+		return -EINVAL;
 
 	__raw_writew(PBC_BCTRL3_FSH_MOD, PBC_BCTRL3_CLEAR_REG);
 	__raw_writew(PBC_BCTRL3_FSH_VBUS_EN, PBC_BCTRL3_CLEAR_REG);
-
 	usbh1_hs_active = 1;
+	return 0;
 }
 void gpio_usbh1_inactive(void)
 {
@@ -227,44 +229,46 @@ void gpio_usbh1_inactive(void)
 	usbh1_hs_active = 0;
 }
 
-static int usbh2_hs_active = 0;
+static int usbh2_hs_active;
 /*
- * conflicts with CSPI1 (Atlas) and CSPI2 (Connector)
+ * conflicts with CSPI1 (MC13783) and CSPI2 (Connector)
  */
-void gpio_usbh2_active(void)
+int gpio_usbh2_active(void)
 {
 	if (usbh2_hs_active)
-		return;
+		return 0;
 
-	gpio_set_puen(MX27_PIN_USBH2_CLK, 0);
-	gpio_set_puen(MX27_PIN_USBH2_DIR, 0);
-	gpio_set_puen(MX27_PIN_USBH2_DATA7, 0);
-	gpio_set_puen(MX27_PIN_USBH2_NXT, 0);
-	gpio_set_puen(MX27_PIN_USBH2_STP, 0);
-	gpio_set_puen(MX27_PIN_CSPI2_SS2, 0);
-	gpio_set_puen(MX27_PIN_CSPI2_SS1, 0);
-	gpio_set_puen(MX27_PIN_CSPI2_SS0, 0);
-	gpio_set_puen(MX27_PIN_CSPI2_SCLK, 0);
-	gpio_set_puen(MX27_PIN_CSPI2_MISO, 0);
-	gpio_set_puen(MX27_PIN_CSPI2_MOSI, 0);
-	gpio_set_puen(MX27_PIN_CSPI1_SS2, 0);
-
-	gpio_request_mux(MX27_PIN_USBH2_CLK, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBH2_DIR, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBH2_DATA7, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBH2_NXT, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBH2_STP, GPIO_MUX_PRIMARY);
-
-	gpio_request_mux(MX27_PIN_CSPI2_SS2, GPIO_MUX_ALT);
-	gpio_request_mux(MX27_PIN_CSPI2_SS1, GPIO_MUX_ALT);
-	gpio_request_mux(MX27_PIN_CSPI2_SS0, GPIO_MUX_ALT);
-	gpio_request_mux(MX27_PIN_CSPI2_SCLK, GPIO_MUX_ALT);
-	gpio_request_mux(MX27_PIN_CSPI2_MISO, GPIO_MUX_ALT);
-	gpio_request_mux(MX27_PIN_CSPI2_MOSI, GPIO_MUX_ALT);
-	gpio_request_mux(MX27_PIN_CSPI1_SS2, GPIO_MUX_ALT);
-	__raw_writew(PBC_BCTRL3_HSH_EN, PBC_BCTRL3_CLEAR_REG);
+	if (gpio_set_puen(MX27_PIN_USBH2_CLK, 0) ||
+	    gpio_set_puen(MX27_PIN_USBH2_DIR, 0) ||
+	    gpio_set_puen(MX27_PIN_USBH2_DATA7, 0) ||
+	    gpio_set_puen(MX27_PIN_USBH2_NXT, 0) ||
+	    gpio_set_puen(MX27_PIN_USBH2_STP, 0) ||
+	    gpio_set_puen(MX27_PIN_CSPI2_SS2, 0) ||
+	    gpio_set_puen(MX27_PIN_CSPI2_SS1, 0) ||
+	    gpio_set_puen(MX27_PIN_CSPI2_SS0, 0) ||
+	    gpio_set_puen(MX27_PIN_CSPI2_SCLK, 0) ||
+	    gpio_set_puen(MX27_PIN_CSPI2_MISO, 0) ||
+	    gpio_set_puen(MX27_PIN_CSPI2_MOSI, 0) ||
+	    gpio_set_puen(MX27_PIN_CSPI1_SS2, 0))
+		return -EINVAL;
+
+	if (gpio_request_mux(MX27_PIN_USBH2_CLK, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBH2_DIR, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBH2_DATA7, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBH2_NXT, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBH2_STP, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_CSPI2_SS2, GPIO_MUX_ALT) ||
+	    gpio_request_mux(MX27_PIN_CSPI2_SS1, GPIO_MUX_ALT) ||
+	    gpio_request_mux(MX27_PIN_CSPI2_SS0, GPIO_MUX_ALT) ||
+	    gpio_request_mux(MX27_PIN_CSPI2_SCLK, GPIO_MUX_ALT) ||
+	    gpio_request_mux(MX27_PIN_CSPI2_MISO, GPIO_MUX_ALT) ||
+	    gpio_request_mux(MX27_PIN_CSPI2_MOSI, GPIO_MUX_ALT) ||
+	    gpio_request_mux(MX27_PIN_CSPI1_SS2, GPIO_MUX_ALT))
+		return -EINVAL;
 
+	__raw_writew(PBC_BCTRL3_HSH_EN, PBC_BCTRL3_CLEAR_REG);
 	usbh2_hs_active = 1;
+	return 0;
 }
 void gpio_usbh2_inactive(void)
 {
@@ -302,29 +306,32 @@ void gpio_usbh2_inactive(void)
 	usbh2_hs_active = 0;
 }
 
-static int usbotg_hs_active = 0;
+static int usbotg_hs_active;
 int gpio_usbotg_hs_active(void)
 {
 	if (usbotg_hs_active)
 		return 0;
 
-	gpio_request_mux(MX27_PIN_USBOTG_DATA5, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBOTG_DATA6, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBOTG_DATA0, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBOTG_DATA2, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBOTG_DATA1, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBOTG_DATA3, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBOTG_DATA4, GPIO_MUX_PRIMARY);
-
-	gpio_request_mux(MX27_PIN_USBOTG_DIR, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBOTG_STP, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBOTG_NXT, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBOTG_CLK, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USBOTG_DATA7, GPIO_MUX_PRIMARY);
-
-	gpio_request_mux(MX27_PIN_USB_OC_B, GPIO_MUX_PRIMARY);
-	gpio_request_mux(MX27_PIN_USB_PWR, GPIO_MUX_PRIMARY);
+	if (gpio_request_mux(MX27_PIN_USBOTG_DATA5, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBOTG_DATA6, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBOTG_DATA0, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBOTG_DATA2, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBOTG_DATA1, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBOTG_DATA3, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBOTG_DATA4, GPIO_MUX_PRIMARY) ||
+
+	    gpio_request_mux(MX27_PIN_USBOTG_DIR, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBOTG_STP, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBOTG_NXT, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBOTG_CLK, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USBOTG_DATA7, GPIO_MUX_PRIMARY) ||
+
+	    gpio_request_mux(MX27_PIN_USB_OC_B, GPIO_MUX_PRIMARY) ||
+	    gpio_request_mux(MX27_PIN_USB_PWR, GPIO_MUX_PRIMARY))
+		return -EINVAL;
+
 	__raw_writew(PBC_BCTRL3_OTG_HS_EN, PBC_BCTRL3_CLEAR_REG);
+	__raw_writew(PBC_BCTRL3_OTG_VBUS_EN, PBC_BCTRL3_CLEAR_REG);
 
 	usbotg_hs_active = 1;
 	return 0;
@@ -363,7 +370,7 @@ int gpio_usbotg_fs_active(void)
 
 void gpio_usbotg_fs_inactive(void)
 {
-	return gpio_usbotg_hs_inactive();
+	gpio_usbotg_hs_inactive();
 }
 
 /*!
-- 
1.5.4.4

